### PR TITLE
Enhance the artifact upload plugin to allow for compressing and extracting artifacts

### PR DESCRIPTION
**Pull request by thien-nguyenthanh**
- Todo: Add bats test

### CODE REVIEW

File: hooks/pre-command [hunk 1]
```diff
@@ -80,8 +86,14 @@ elif [[ "${MULTIPLE_DOWNLOADS}" == "true" ]]; then
           if ! [[ -d $(dirname "${!dest_env_var}") ]]; then
             mkdir -p "$(dirname "${!dest_env_var}")"
           fi
-          echo "~~~ Moving [${source}] to [${!dest_env_var}]..."
-          mv "${source}" "${!dest_env_var}"
+          # Handler for extract artifacts
+          if [[ "${BUILDKITE_PLUGIN_ARTIFACTS_DOWNLOAD_EXTRACT:-false}" =~ ^(true|on|1)$ ]] ; then
```
Comment (chrisfowles) [inline hooks/pre-command:90]: for multiple downloads - we need to be looking at the indexed parameter; similar to the way that `source_env_var` is being handled.

File: hooks/post-command [hunk 2]
```diff
@@ -86,9 +92,12 @@ elif [[ "${MULTIPLE_UPLOADS}" == "true" ]]; then
         source_env_var="BUILDKITE_PLUGIN_ARTIFACTS_UPLOAD_${index}_FROM"
         dest_env_var="BUILDKITE_PLUGIN_ARTIFACTS_UPLOAD_${index}_TO"
         if [[ -n "${!source_env_var:-}" ]] && [[ -n "${!dest_env_var:-}" ]]; then
-          if [[ -e ${!source_env_var} ]]; then
+          if [[ -e ${!source_env_var} && !"${BUILDKITE_PLUGIN_ARTIFACTS_DOWNLOAD_COMPRESS:-false}" =~ ^(true|on|1)$]]; then
```
Comment (chrisfowles) [inline hooks/post-command:95]: same as with downloads, we need to be looking at indexed params for this

**Review by formerly-chrisfowles — CHANGES_REQUESTED** [review]
looking good - just a couple of fixes around when we're handling multiple downloads/uploads

File: hooks/pre-command (hunk 1 above)
Comment (thien-enouvo) [reply hooks/pre-command]: Yeah, I fixed it.

**Review by formerly-chrisfowles — APPROVED** [review]

File: tests/command.bats [hunk 3]
```diff
@@ -211,6 +292,27 @@ load '/usr/local/lib/bats/load.bash'
   unset BUILDKITE_PLUGIN_ARTIFACTS_DOWNLOAD_TO
 }
 
+@test "Post-command uploads artifacts with a compressing option" {
+  mkdir /tmp/data && touch /tmp/data/foo.log
```
Comment (chrisfowles) [inline tests/command.bats:296]: some of these tests are failing - i think you'll need to clean up your temp directories after each test as they're conflicting with each other.

**Review by formerly-chrisfowles — CHANGES_REQUESTED** [review]
some of your tests are failing - you can see which ones by running docker-compose up.

I think it's mainly just that you need to clean up after each individual test

File: tests/command.bats (hunk 3 above)
Comment (thien-enouvo) [reply tests/command.bats]: @chrisfowles I have fixed it with `mkdir `option ignore error if existing.
I can run test successfully.

Comment (chrisfowles) [reply tests/command.bats]: We should still clean up after ourselves to ensure that we're not impacting any other downstream tests in the future

Comment (thien-enouvo) [reply tests/command.bats]: @chrisfowles ok Chris, help me squash commit  before merged to master.

**Review by formerly-chrisfowles — APPROVED** [review]

**Comment by formerly-chrisfowles** [comment]
great job - looks good.

**Pull request merged**